rosetta-source/src/main/rosetta/product-qualification-func.rosetta

```diff
@@ -261,34 +277,38 @@ func Qualify_CreditDefaultSwaption: <"This product qualification is temporary un
+                    economicTerms -> payout -> optionPayout only-element -> optionUnderlier -> Product ->> economicTerms
                 ) = True
 
-func Qualify_UnderlierProduct_Equity: <"Qualifies a product as having the Asset Class classification Equity.">
+func Qualify_UnderlierObservable_Equity: <"Qualifies an Observable as having the Asset Class classification Equity.">
     inputs:
-        underlier Product (1..1)
+        underlier Observable (1..1)
     output:
         is_product boolean (1..1)
 
     set is_product:
-        underlier -> security -> securityType = SecurityTypeEnum -> Equity
-            or (underlier -> security -> securityType = SecurityTypeEnum -> Fund
-                and underlier -> security -> fundType = FundProductTypeEnum -> ExchangeTradedFund)
-            or (underlier -> security -> securityType = SecurityTypeEnum -> Fund
-                and underlier -> security -> fundType = FundProductTypeEnum -> MutualFund)
-            or underlier -> security -> securityType = SecurityTypeEnum -> Warrant
-            or underlier -> index ->> assetClass = AssetClassEnum -> Equity
+        Qualify_SecurityTypeEquity(underlier -> Asset -> Instrument -> Security)
+            or underlier -> Index ->> assetClass = AssetClassEnum -> Equity
             // Qualifies that the underlier is a basket composed of equity products only
-            or (underlier -> basket exists
-                and (underlier -> basket -> basketConstituent -> asset -> Instrument -> Security -> securityType any = SecurityTypeEnum -> Equity
-                    or (underlier -> basket -> basketConstituent -> asset -> Instrument -> Security -> securityType any = SecurityTypeEnum -> Fund
-                        and underlier -> basket -> basketConstituent -> asset -> Instrument -> Security -> fundType any = FundProductTypeEnum -> ExchangeTradedFund)
-                    or (underlier -> basket -> basketConstituent -> asset -> Instrument -> Security -> securityType any = SecurityTypeEnum -> Fund
-                        and underlier -> basket -> basketConstituent -> asset -> Instrument -> Security -> fundType any = FundProductTypeEnum -> MutualFund)
-                    or underlier -> basket -> basketConstituent -> asset -> Instrument -> Security -> securityType any = SecurityTypeEnum -> Warrant
-                    or underlier -> basket -> basketConstituent -> index ->> assetClass any = AssetClassEnum -> Equity 
+            or (underlier -> Basket exists
+                and (Qualify_SecurityTypeEquity(underlier -> Basket -> basketConstituent -> Asset -> Instrument -> Security)
+                    or underlier -> Basket -> basketConstituent -> Index ->> assetClass all = AssetClassEnum -> Equity 
                     ))
 
+func Qualify_SecurityTypeEquity:  <"Qualifies that the security type, for all of the provided securities, is Equity-related.">
+    inputs:
+        security Security (1..*)
```

Comment:
Rewrite logic using single cardinality for `security` instead, see other comment.

Reply:
done